tee-worker/omni-executor/rpc-server/src/methods/omni/get_approve_builder_fee_data.rs

```diff
@@ -0,0 +1,286 @@
+use crate::{
+	error_code::*, server::RpcContext, verify_auth::verify_email_authentication, ErrorCode,
+};
+use alloy::{
+	dyn_abi::Eip712Domain,
+	primitives::{keccak256, Address, B256},
+	sol_types::{eip712_domain, SolValue},
+};
+use chrono::Utc;
+use executor_primitives::{utils::hex::hex_encode, ChainId, Identity, UserAuth, UserId};
+use jsonrpsee::{types::ErrorObject, RpcModule};
+use pumpx::pubkey_to_address;
+use serde::{Deserialize, Serialize, Serializer};
+use signer_client::ChainType;
+use std::{convert::TryFrom, str::FromStr};
+use tracing::error;
+
+#[derive(Debug, Deserialize)]
+pub struct GetApproveBuilderFeeDataParams {
+	pub user_id: UserId,
+	pub user_auth: UserAuth,
+	pub client_id: String,
+	pub max_fee_rate: String,
+	pub builder: String,
+	pub chain_id: ChainId,
+}
+
+#[derive(Serialize, Clone)]
+pub struct GetApproveBuilderFeeDataResponse {
+	pub main_address: String,
+	pub approve_builder_fee_data: ApproveBuilderFeeData,
+}
+
+#[derive(Serialize, Clone)]
+pub struct ApproveBuilderFeeData {
+	pub action: ApproveBuilderFeeAction,
+	pub nonce: u64,
+	pub signature: String,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct ApproveBuilderFeeAction {
+	#[serde(serialize_with = "serialize_hex")]
+	pub signature_chain_id: u64,
+	pub hyperliquid_chain: String,
+	pub max_fee_rate: String,
+	pub builder: Address,
+	pub nonce: u64,
+}
+
+fn is_testnet_chain(chain_id: ChainId) -> bool {
+	match chain_id {
+		// Testnet chain IDs
+		11155111 => true, // Ethereum Sepolia
+		97 => true,       // BSC Testnet
+		80001 => true,    // Polygon Mumbai
+		421614 => true,   // Arbitrum Sepolia
+		998 => true,      // HyperEVM Testnet
+		1337 => true,     // Local Anvil
+		31337 => true,    // Local Anvil
+		// Mainnet chain IDs
+		1 => false,     // Ethereum Mainnet
+		56 => false,    // BSC Mainnet
+		137 => false,   // Polygon Mainnet
+		42161 => false, // Arbitrum Mainnet
+		999 => false,   // HyperEVM Mainnet
+		_ => {
+			// Default to mainnet for unknown chain IDs
+			false
+		},
+	}
+}
+
+impl ApproveBuilderFeeAction {
```

Comment:
Do you consider moving these into a Trait?

They seem to be similar except for `struct_hash()`